### PR TITLE
Async malicious: Does not work

### PR DESCRIPTION
**Pull request by richajaindce**

### CODE REVIEW

File: src/secret_sharing/malicious_replicated.rs [hunk 1]
```diff
@@ -140,38 +146,50 @@ impl<F: Field> Mul<F> for MaliciousReplicated<F> {
...
-        UnauthorizedDowngradeWrapper((
-            self.0.downgrade().access_without_downgrade(),
-            self.1.downgrade().access_without_downgrade(),
-        ))
+    type Target = (<T>::Target, <U>::Target);
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target>
+    where
+        Self::Target: Send,
+    {
+        let one = self.0.downgrade().await.access_without_downgrade();
```
Comment (richajaindce) [inline src/secret_sharing/malicious_replicated.rs]: self.0 is giving error : `<T as Downgrade>::Target` cannot be sent between threads safely
the trait `std::marker::Send` is not implemented for `<T as Downgrade>::Target`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message?3#file%3A%2F%2F%2FUsers%2Frichaj%2FProposals%2FIPA%2Fraw-ipa%2Fsrc%2Fsecret_sharing%2Fmalicious_replicated.rs)
malicious_replicated.rs(171, 26): required by a bound introduced by this call
malicious_replicated.rs(27, 23): required by a bound in `Downgrade::downgrade`
malicious_replicated.rs(167, 23): consider further restricting the associated type: `, <T as Downgrade>::Target: std::marker::Send`

Comment (martinthomson) [reply src/secret_sharing/malicious_replicated.rs]: join()?

Comment (richajaindce) [reply src/secret_sharing/malicious_replicated.rs]: This is still not happy, downgrade is async and access_without_downgrade is where we join
`<U as Downgrade>::Target` is not a future
within `futures::future::Join<<T as Downgrade>::Target, <U as Downgrade>::Target>`, the trait `futures::Future` is not implemented for `<U as Downgrade>::Target`
<U as Downgrade>::Target must be a future or must implement `IntoFuture` to be awaited
the return type of a function must have a statically known size

File: src/secret_sharing/malicious_replicated.rs [hunk 2]
```diff
@@ -140,38 +144,45 @@ impl<F: Field> Mul<F> for MaliciousReplicated<F> {
...
-            self.0.downgrade().access_without_downgrade(),
-            self.1.downgrade().access_without_downgrade(),
-        ))
+    type Target = (<T>::Target, <U>::Target);
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target>
+    {
+        let one = self.0.downgrade();
+        let two = self.1.downgrade();
+        let output = join(one, two).await;
+        UnauthorizedDowngradeWrapper((output.0.access_without_downgrade(), output.1.access_without_downgrade()))
```
Comment (richajaindce) [inline src/secret_sharing/malicious_replicated.rs]: Even this version does not work:
let __self: (T, U)
future cannot be sent between threads safely
within `impl futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>>`, the trait `std::marker::Send` is not implemented for `<T as Downgrade>::Target`
required for the cast from `impl futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>>` to the object type `dyn futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>> + std::marker::Send`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message?3#file%3A%2F%2F%2FUsers%2Frichaj%2FProposals%2FIPA%2Fraw-ipa%2Fsrc%2Fsecret_sharing%2Fmalicious_replicated.rs)
malicious_replicated.rs(167, 36): future is not `Send` as this value is used across an await
malicious_replicated.rs(167, 42): `join(one, two)` is later dropped here

File: src/secret_sharing/malicious_replicated.rs [hunk 3]
```diff
@@ -140,38 +146,42 @@ impl<F: Field> Mul<F> for MaliciousReplicated<F> {
...
-            self.0.downgrade().access_without_downgrade(),
-            self.1.downgrade().access_without_downgrade(),
-        ))
+    type Target = (<T>::Target, <U>::Target);
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+        let one = self.0.downgrade();
+        let two = self.1.downgrade();
+        let out = join(one, two).await;
+        let output = (out.0.access_without_downgrade(), out.1.access_without_downgrade());
+        UnauthorizedDowngradeWrapper(output)
```
Comment (richajaindce) [inline src/secret_sharing/malicious_replicated.rs]: @martinthomson following error is seen:
raw_ipa::secret_sharing::malicious_replicated
pub struct UnauthorizedDowngradeWrapper<T>
[1 implementation](vscode-file://vscode-app/private/var/folders/_r/s7hk62fn5c16yfvqlg3whg400000gn/T/AppTranslocation/71EC7F17-8378-470B-819D-C269580AE45A/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)

future cannot be sent between threads safely
within `impl futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>>`, the trait `std::marker::Send` is not implemented for `<T as Downgrade>::Target`
required for the cast from `impl futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>>` to the object type `dyn futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>> + std::marker::Send`

File: src/secret_sharing/malicious_replicated.rs [hunk 4]
```diff
@@ -140,38 +146,42 @@ impl<F: Field> Mul<F> for MaliciousReplicated<F> {
...
-            self.into_iter()
-                .map(|v| v.downgrade().access_without_downgrade())
-                .collect(),
-        )
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+        let output = join_all(self
+            .into_iter()
+            .map(|v| async move { v.downgrade().await.access_without_downgrade() })
+            .collect::<Vec<_>>()).await;
+        UnauthorizedDowngradeWrapper(output)
```
Comment (richajaindce) [inline src/secret_sharing/malicious_replicated.rs]: @martinthomson following error is seen:
future cannot be sent between threads safely
the trait `std::marker::Send` is not implemented for `<T as Downgrade>::Target`
required for the cast from `impl futures::Future<Output = UnauthorizedDowngradeWrapper<Vec<<T as Downgrade>::Target>>>` to the object type `dyn futures::Future<Output = UnauthorizedDowngradeWrapper<Vec<<T as Downgrade>::Target>>> + std::marker::Send`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message?7#file%3A%2F%2F%2FUsers%2Frichaj%2FProposals%2FIPA%2Fraw-ipa%2Fsrc%2Fsecret_sharing%2Fmalicious_replicated.rs)
malicious_replicated.rs(183, 34): future is not `Send` as this value is used across an await
malicious_replicated.rs(183, 40): the value is later dropped here

**Comment by richajaindce** [comment]
This is resolved and now in review! :)

**Pull request closed without merging**